### PR TITLE
194 task collection api

### PR DESCRIPTION
**Pull request by jacopo-exact**

### CODE REVIEW

File: fractal_server/utils.py [hunk 1]
```diff
@@ -74,6 +100,8 @@ async def execute_command(*, cwd: Path, command: str) -> str:
     command_split = shlex_split(command)
     cmd, *args = command_split
 
+    logger = set_logger(logger_name=logger_name)
```
Comment (tcompa) [inline fractal_server/utils.py:103]: Should you also handle `logger_name=None`?

File: fractal_server/utils.py [hunk 2]
```diff
@@ -82,6 +110,8 @@ async def execute_command(*, cwd: Path, command: str) -> str:
         cwd=cwd,
     )
     stdout, stderr = await proc.communicate()
+    logger = logging.getLogger(logger_name)
```
Comment (tcompa) [inline fractal_server/utils.py:113]: Isn't `logger` the same that you already obtained with `set_logger`?

File: fractal_server/app/api/v1/task.py [hunk 3]
```diff
@@ -51,27 +47,42 @@ def get_task(
...
+    await asyncio.gather(*[db.refresh(t) for t in task_db_list])
+    return task_db_list
+
+
+@router.post(
+    "/pip/",
+    response_model=List[TaskRead],
+    status_code=status.HTTP_201_CREATED,
+)
+async def create_task_pip(
```
Comment (tcompa) [inline fractal_server/app/api/v1/task.py:67]: Minor comments: wouldn't we name it `create_tasks_pip` (plural)?

File: fractal_server/app/api/v1/task.py [hunk 4]
```diff
@@ -51,27 +47,42 @@ def get_task(
...
+    task_collect: TaskCollectPip,
+    collection_type: str = "pip",
+    global_task: bool = True,
     user: User = Depends(current_active_user),
```
Comment (tcompa) [inline fractal_server/app/api/v1/task.py:71]: Possibly trivial question: what does it mean to have `user` as an argument, if it is not used in the endpoint function?

Comment (tcompa) [reply fractal_server/app/api/v1/task.py:71]: My best guess: later on, we'll want to associate some tasks to projects, and projects may be access-limited to certain users. Is this the reason?

File: tests/test_task_api.py [hunk 5]
```diff
@@ -17,22 +15,26 @@ async def test_task_get_list(db, client, task_factory, MockCurrentUser):
...
-    task = TaskCreate(
-        name="mytask",
-        command="cmd",
-        source="/source",
-        input_type="Any",
-        output_type="Any",
+    PREFIX = "/api/v1/task"
+
+    task_collection = dict(
+        collection_type="pypi", package=dummy_task_package.as_posix()
```
Comment (tcompa) [inline tests/test_task_api.py:28]: Is `collection_type` meant to be `pypi` or `pip`?

File: fractal_server/app/api/v1/task.py [hunk 6]
```diff
@@ -51,27 +47,42 @@ def get_task(
...
+
+
+@router.post(
+    "/pip/",
+    response_model=List[TaskRead],
+    status_code=status.HTTP_201_CREATED,
+)
+async def create_task_pip(
+    task_collect: TaskCollectPip,
+    collection_type: str = "pip",
```
Comment (tcompa) [inline fractal_server/app/api/v1/task.py:69]: `collection_type` is never used

File: tests/test_task_api.py (hunk 5 above)
Comment (tcompa) [reply tests/test_task_api.py:28]: (by the way, this argument of `create_task_pip` is currently never used.. maybe it was some leftover from past code or a placeholder for future features)

File: fractal_server/utils.py (hunk 1 above)
Comment (jacopo-exact) [reply fractal_server/utils.py:103]: when None, `getLogger` returns the default logger.

File: fractal_server/utils.py (hunk 2 above)
Comment (jacopo-exact) [reply fractal_server/utils.py:113]: you are right. will fix

File: tests/fixtures_tasks.py [hunk 7]
```diff
@@ -39,3 +43,42 @@ def assemble_args(self, extra: Dict[str, Any] = None):
             full_args.update(extra)
         full_args.update(self.arguments)
         return full_args
+
+
+async def execute_command(cmd, **kwargs):
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply tests/fixtures_tasks.py:48]: in `fractal_server/utils.py` we use [`asyncio.create_subprocess_exec`](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.create_subprocess_exec), which executes a command without access to an actual shell (safer and more performant). Here I'm using [`asyncio.create_subprocess_shell`](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.create_subprocess_shell), which opens an actual shell, i.e., emulates what a user would actually do from command line.

File: fractal_server/tasks/collection.py [hunk 8]
```diff
@@ -17,47 +17,111 @@
...
+        Checks if package corresponds to an existing path on the filesystem
+
+        In this case, the user is providing directly a package file, rather
+        than a remote one from PyPI. We set the `package_path` attribute and
+        get the actual package name and version from the package file name.
+        """
+        package_path = Path(values["package"])
+        if package_path.exists():
+            values["package_path"] = package_path
+            values["package"], values["version"], *_ = package_path.name.split(
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply fractal_server/tasks/collection.py:56]: You are right: here we are **only** dealing with `pip` installable packages. Isolated packages will need a separate collection subsystem

File: fractal_server/tasks/collection.py [hunk 9]
```diff
@@ -17,47 +17,111 @@
...
+
+    @property
+    def source(self):
+        if self.is_local_package:
+            return f"pip-local:{self.package_path.name}"
+        else:
+            return f"pip:{self.package}=={self.version}"
+
+
+def _package_from_path(wheel_path: Path) -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply fractal_server/tasks/collection.py:69]: yes you are right

File: fractal_server/tasks/collection.py [hunk 10]
```diff
@@ -157,13 +219,25 @@ async def _pip_install(
...
+        version_string = f"=={task_pkg.version}" if task_pkg.version else ""
+        extras = (
+            f"[{task_pkg.package_extras}]" if task_pkg.package_extras else ""
+        )
+        pip_install_str = f"{task_pkg.package}{extras}{version_string}"
 
-    await execute_command(cwd=venv_path, command=cmd_install)
-    stdout_inspect = await execute_command(cwd=venv_path, command=cmd_inspect)
+    cmd_install = f"{pip} install {pip_install_str}"
+    cmd_inspect = f"{pip} show -f {task_pkg.package}"
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply fractal_server/tasks/collection.py:233]: for ease of parsing

File: fractal_server/app/api/v1/task.py (hunk 6 above)
Comment (jacopo-exact) [reply fractal_server/app/api/v1/task.py:69]: correct, will remove

File: fractal_server/tasks/collection.py [hunk 11]
```diff
@@ -17,47 +17,111 @@
...
+
+    package_dir = f"{task_pkg.package}{task_pkg.version or ''}"
     if settings.FRACTAL_ROOT:
+        # It should be always true, we are only checking that the system is
+        # fully configured
         env_path = settings.FRACTAL_ROOT / user / package_dir
     # TODO check the access right of the env_path and subdirs
-    env_path.mkdir(exist_ok=True, parents=True)
+    env_path.mkdir(exist_ok=False, parents=True)
+    return env_path
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply fractal_server/tasks/collection.py:99]: will refactor to `venv`

File: tests/test_task_api.py (hunk 5 above)
Comment (jacopo-exact) [reply tests/test_task_api.py:28]: correct, will remove

File: fractal_server/tasks/collection.py [hunk 12]
```diff
@@ -17,47 +17,111 @@
...
+    **_,
+) -> Path:
     settings = Inject(get_settings)
     # assemble installation path
-    package_dir = f"{package}{version or ''}"
+    package_path = Path(task_pkg.package)
+    if package_path.exists():
+        # The package is a local package in the filesystem rather than a
+        # remote PyPI package. Extract package and version from basename
+        package, version = _package_from_path(package_path)
```
[the remark this answers is not in the document]
Comment (jacopo-exact) [reply fractal_server/tasks/collection.py:90]: you are right, and even better, we can remove this part altogether as it is already taken care of by the root validator.

**Pull request merged**